Java/core/src/main/java/com/amazon/randomcutforest/RandomCutForest.java

```diff
@@ -179,62 +180,52 @@
-        });
-        checkArgument(builder.dimensions > 0, "dimensions must be greater than 0");
-        builder.lambda.ifPresent(lambda -> {
-            checkArgument(lambda >= 0, "lambda must be greater than or equal to 0");
-        });
-        builder.threadPoolSize.ifPresent(n -> checkArgument((n > 0) || ((n == 0) && !builder.parallelExecutionEnabled),
-                "threadPoolSize must be greater/equal than 0. To disable thread pool, set parallel execution to 'false'."));
+    protected <Q> RandomCutForest(Builder<?> builder, IUpdateCoordinator<Q> updateCoordinator,
+            ComponentList<Q> componentModels) {
+        this(builder);
 
-        numberOfTrees = builder.numberOfTrees;
-        sampleSize = builder.sampleSize;
-        outputAfter = builder.outputAfter.orElse((int) (sampleSize * DEFAULT_OUTPUT_AFTER_FRACTION));
-        dimensions = builder.dimensions;
-        lambda = builder.lambda.orElse(1.0 / (DEFAULT_SAMPLE_SIZE_COEFFICIENT_IN_LAMBDA * sampleSize));
-        storeSequenceIndexesEnabled = builder.storeSequenceIndexesEnabled;
-        centerOfMassEnabled = builder.centerOfMassEnabled;
-        parallelExecutionEnabled = builder.parallelExecutionEnabled;
-        compactEnabled = builder.compactEnabled;
-        saveTreeData = builder.saveTreeData;
-        ArrayList<SamplerPlusTree> components = new ArrayList<>(numberOfTrees);
+        if (updateCoordinator != null) {
+            checkArgument(componentModels != null,
+                    "updateCoordinator and componentModels most both be null or nonnull");
+            this.updateCoordinator = updateCoordinator;
+            this.componentModels = componentModels;
+        } else {
+            checkArgument(componentModels == null,
+                    "updateCoordinator and componentModels most both be null or nonnull");
+            if (compactEnabled) {
+                initCompact();
+            } else {
+                initNoncompact();
+            }
+        }
 
-        // TODO this will replace `components` in a future revision
-        componentModels = new ArrayList<>(numberOfTrees);
-
-        // If a random seed was given, use it to create a new Random. Otherwise, call
-        // the 0-argument constructor
-        rng = builder.randomSeed.map(Random::new).orElseGet(Random::new);
+        // TODO the rest of this constructor will be updated in a future revision
 
         if (compactEnabled) {
             pointStore = new PointStoreDouble(dimensions, sampleSize * numberOfTrees + 1);
-            updateCoordinator = new PointStoreCoordinator(pointStore);
+            this.updateCoordinator = new PointStoreCoordinator(pointStore);
         } else {
             pointStore = null;
-            updateCoordinator = new PointSequencer();
+            this.updateCoordinator = new PassThroughCoordinator();
         }
 
+        ArrayList<SamplerPlusTree> components = new ArrayList<>(numberOfTrees);
```

Comment:
issues? what happen to trees created in line 211, 213?

Reply:
Currently nothing. In the next PR I'll update the executor classes so that they take in a ComponentList argument (currently they expect a List<SamplerPlusTree>), and that would made this PR bigger and more complicated.